### PR TITLE
Clarify registration/authentication instructions in README

### PR DESCRIPTION
**Pull request by LINKIWI**
Hi, thanks for creating this module! I've used it to implement U2F authentication in one of my [personal projects](https://github.com/LINKIWI/apache-auth) but had some trouble understanding the code examples in the README instructions to get everything working on both the server and client.

This PR clarifies the U2F registration/authentication instructions in the README, especially clarifying points of confusion I encountered as someone who is not familiar with the official U2F API spec.

I have also fixed an error on the client-side documentation. Namely, the existing `u2f` client calls are inconsistent with what the [official FIDO client-side API spec](https://fidoalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-javascript-api.html) specifies as input to `window.u2f.register` and `window.u2f.sign`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
 }
 ```
 
+##### Client logic:
+
+```javascript
+const authReq = ...;  // Retrieve this from hitting the authentication challenge endpoint
+const {appId, challenge, version, keyHandle} = authReq;
+
+window.u2f.sign(appId, challenge, [{version, keyHandle}], (authRes) => {
```
Comment (LINKIWI) [inline README.md]: The input parameters to `u2f.sign` are fixed here relative to what's already in the README. Reference: [official Javascript U2F client API implementation](https://demo.yubico.com/js/u2f-api.js) and [FIDO API spec](https://fidoalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-javascript-api.html)

Relevant snippet:

```
/**
 * Dispatches an array of sign requests to available U2F tokens.
 * If the JS API version supported by the extension is unknown, it first sends a
 * message to the extension to find out the supported API version and then it sends
 * the sign request.
 * @param {string=} appId
 * @param {string=} challenge
 * @param {Array<u2f.RegisteredKey>} registeredKeys
 * @param {function((u2f.Error|u2f.SignResponse))} callback
 * @param {number=} opt_timeoutSeconds
 */
u2f.sign = function(appId, challenge, registeredKeys, callback, opt_timeoutSeconds) {
```

File: README.md [hunk 2]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
 
+##### Client logic:
+
+Note that the `window.u2f` object is defined in the official [Javascript U2F API](https://github.com/google/u2f-ref-code), for which a polyfill is [available as an npm module](https://www.npmjs.com/package/u2f-api-polyfill).
+
+```javascript
+const registrationReq = ...  // Retrieve this from hitting the registration challenge endpoint
+const {appId, version, challenge} = registrationReq;
+
+window.u2f.register(appId, [{version, challenge}], [], (registrationRes) => {
```
Comment (LINKIWI) [inline README.md]: The input parameters to `u2f.register` are fixed here relative to what's already in the README. Reference: [official Javascript U2F client API implementation](https://demo.yubico.com/js/u2f-api.js) and [FIDO API spec](https://fidoalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-javascript-api.html)

Relevant snippet:

```
/**
 * Dispatches register requests to available U2F tokens. An array of sign
 * requests identifies already registered tokens.
 * If the JS API version supported by the extension is unknown, it first sends a
 * message to the extension to find out the supported API version and then it sends
 * the register request.
 * @param {string=} appId
 * @param {Array<u2f.RegisterRequest>} registerRequests
 * @param {Array<u2f.RegisteredKey>} registeredKeys
 * @param {function((u2f.Error|u2f.RegisterResponse))} callback
 * @param {number=} opt_timeoutSeconds
 */
u2f.register = function(appId, registerRequests, registeredKeys, callback, opt_timeoutSeconds) {
```

File: README.md [hunk 3]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
 
+##### Server endpoint (registration challenge):
+
 ```javascript
-// 1. Check that the user is logged in.
+const u2f = require('u2f');
 
-// 2. (On server) generate user registration request and save it in session:
-var u2f = require('u2f');
+function handler(req, res) {
```
Comment (ashtuchkin) [inline README.md]: I like how you structured it!  Wdyt about merging these two pieces of server code and just using 2 handler functions - one for registration request, another for verification?

File: README.md [hunk 4]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
+  // 1. Check that the user is logged in.
 
-var req = u2f.request("<appId>");
-session.authRequest = req;
+  // 2. Generate a registration request and save it in the session.
+  // The app ID is usually the fully qualified URL of your website. The website MUST be HTTPS, or
+  // else the registration will fail client-side.
```
Comment (ashtuchkin) [inline README.md]: else -> otherwise

File: README.md [hunk 5]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
 
-// 2. (On server) generate user registration request and save it in session:
-var u2f = require('u2f');
+function handler(req, res) {
+  // 1. Check that the user is logged in.
 
-var req = u2f.request("<appId>");
-session.authRequest = req;
+  // 2. Generate a registration request and save it in the session.
+  // The app ID is usually the fully qualified URL of your website. The website MUST be HTTPS, or
```
Comment (ashtuchkin) [inline README.md]: Maybe move the app_id to be a global const for the app and move this comment there?

File: README.md [hunk 6]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
+  // else the registration will fail client-side.
+  const registrationReq = u2f.request('<appId>');
+  req.session.registrationReq = registrationReq;
 
-// 3. Send req to the client and use U2F API to request registration, then send result back.
-// (On client)  u2f.register([req], [], function(res) { /* send res back to server */ })
+  // 3. Send the registration request to the client, who will use the Javascript U2F API to sign
+  // the registration request, and send it back to the server for verification.
```
Comment (ashtuchkin) [inline README.md]: It might make sense to add that registration request is a json object.

File: README.md [hunk 7]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
+function handler(req, res) {
+  // 1. Check that the user is logged in.
 
-var req = u2f.request("<appId>");
-session.authRequest = req;
+  // 2. Generate a registration request and save it in the session.
+  // The app ID is usually the fully qualified URL of your website. The website MUST be HTTPS, or
+  // else the registration will fail client-side.
+  const registrationReq = u2f.request('<appId>');
+  req.session.registrationReq = registrationReq;
```
Comment (ashtuchkin) [inline README.md]: Can you use full names here and below, i.e. `req.session.registrationRequest`

File: README.md (hunk 2 above)
Comment (ashtuchkin) [inline README.md]: Yeah, the interface changed from version 1.0. Thanks for noticing this!

Comment (ashtuchkin) [inline README.md]: I think it also might be better to hide what's inside registrationRequest as it's implementation details:
```javascript
const registrationRequest = ...  // Retrieve this from hitting the registration challenge endpoint

window.u2f.register(registrationRequest.appId, [registrationRequest], [], (registrationRes) => ...
```

File: README.md [hunk 8]
```diff
@@ -22,50 +22,118 @@ although basic usage is shown below.
...
 ### User Authentication Flow
+
+##### Server endpoint (authentication challenge):
+
 ```javascript
-// 1. Check that the user is logged in using password authentication (or some other way).
+const u2f = require('u2f');
+
+function handler(req, res) {
+  // 1. Check that the user is logged in.
```
Comment (ashtuchkin) [inline README.md]: Could you add something like `using password authentication` to make it clearer?

File: README.md (hunk 2 above)
Comment (ashtuchkin) [inline README.md]: Same for authentication flow below

File: README.md (hunk 3 above)
Comment (LINKIWI) [inline README.md]: Good idea - I've combined the two handlers into the same code block, naming one `registrationChallengeHandler` and the other `registrationVerificationHandler`.

File: README.md [hunk 9]
```diff
@@ -22,50 +22,108 @@ although basic usage is shown below.
...
+  // the authentication request, and send it back to the server for verification.
+  return res.send(authRequest);
+}
 
-// 4. (On server) check signature using publicKey from user account.
-var checkres = u2f.checkSignature(session.authRequest, res, user.publicKey);
+function authenticationVerificationHandler(req, res) {
+  // 4. Verify the authentication response from the client against the authentication request saved
+  // in the server-side session.
+  const result = u2f.checkSignature(req.session.authRequest, req.body.authResponse);
```
Comment (jessereitsma) [inline README.md]: For `checkSignature()` the third argument (user's publicKey) is missing.
I got a buffer error, after looking at the code I saw that there are three arguments required.

Comment (LINKIWI) [inline README.md]: Good catch; fixed.

**Comment by ashtuchkin** [comment]
Looks great, thank you!

**Pull request merged**